src/librustc_typeck/check/regionck.rs

```diff
@@ -759,6 +759,11 @@ impl<'a, 'gcx, 'tcx, 'v> Visitor<'v> for RegionCtxt<'a, 'gcx, 'tcx> {
             }
 
             hir::ExprRet(Some(ref ret_expr)) => {
+                if self.call_site_scope.is_none() {
+                    struct_span_err!(self.tcx.sess, expr.span, E0571,
+                                     "return statement cannot be out of a function scope").emit();
```

Comment:
Just have to move it to `ExprRet` type-checking, should be straight-forward (except for the condition not being region-related).